README.md
# MarloweScan

MarloweScan is a tool that allows you to explore on-chain Marlowe contracts and watch their execution in Marlowe terms. This document provides instructions on how to build and develop the MarloweScan tool.

## Using Nix

This repo provides a `shell.nix` file that can be used for developing and building MarloweScan without having to worry about dependencies.

### Building with Nix

To build MarloweScan with Nix, you can use the `nix-build` command with `shell.nix`:

```bash
nix-build shell.nix
```

The resulting executable will be made available in `result/bin/marlowe-scan-exe`

### Development shell

As usual, you can use the `nix-shell` command to enter the development environment from the root folder of the project:

```bash
nix-shell
```

That will make the required Haskell libraries, tools like cabal, stack, the Haskell Language Server, and the `gen-hie` command available in the command line.

### Caching dependencies

In order to take advantage of caching, you can use the `https://nixos.org/channels/nixos-22.11` channel. If using a different channel like `unstable`, building may take a while.

## Using Stack

To build MarloweScan with Stack you just need to write:

```bash
stack build
```

To run MarloweScan with Stack, you can use the following command:

```bash
stack exec marlowe-scan-exe
```

Both these things can be done from within the development environment, and it will set up Stack for you.

## Using Cabal

To build MarloweScan with Cabal, you can use the following command:

```bash
cabal build
```

To run MarloweScan with Cabal, you can use the following command:

```bash
cabal run
```

Both these things can be done from within the development environment, and it will set up Cabal for you.

## Updating hie.yaml

If you add or move or rename Haskell modules, you will need to update the `hie.yaml` file, which is used by the Haskell Language Server (HLS), you can use the following command from within the Nix shell and the `marlowe-scan` root folder:

```bash
fix-hie
```

From outside of the development shell, it can be updated using the following command instead:

```bash
gen-hie --stack > hie.yaml
```

Please note that `hie.yaml` files located outside of the project folder (in ancestor folders) may interfere with HLS.

## Executable parameters

MarloweScan provides the following flags that can be used to configure its behavior:

- `--title-label TEXT`: The label to be shown together with the title in the MarloweScan in parenthesis. (It can be used to display the name of the network that MarloweScan is deployed to.) The default value is `Preprod`.

- `--marlowe-scan-port PORT`: The port number to use for the MarloweScan server. The default value is `8081`.
    
- `--runtime-host HOSTNAME-OR-IP`: The hostname or IP address of the running Marlowe Runtime server. The default value is `builder`.
    
- `--runtime-port PORT`: The port number of the running Marlowe Runtime server. The default value is `8080`.
    
- `--block-explorer HOST`: The hostname or IP address for exploring Cardano blockchain addresses, transactions, etc. The default value is "preprod.cardanoscan.io".

These flags can be set by using the command line when starting the MarloweScan server. For example, to start the server on port `9000` and connect to a runtime server running `Preview` on the IP address `192.168.0.1` and port `8888`, you could use the following command:

```bash
result/bin/marlowe-scan-exe --title-label "Preview" --marlowe-scan-port 9000 --runtime-host 192.168.0.1 --runtime-port 8888 --block-explorer "preview.cardanoscan.io"
```

If you have built the project using Stack you can pass the parameters by adding `--` between the command and the parameters, like this:

```bash
stack exec marlowe-scan-exe -- --title-label "Preview" --marlowe-scan-port 9000 --runtime-host 192.168.0.1 --runtime-port 8888 --block-explorer "preview.cardanoscan.io"
```




CHANGELOG.md
# Changelog for `marlowe-scan`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Opts.hs
{-# LANGUAGE DeriveGeneric #-}

module Opts
  ( TitleLabel (..)
  , BlockExplorerHost (..)
  , MarloweScan (..)
  , RuntimeHost (..)
  , RuntimePort (..)
  , Options (..)
  , mkUrlPrefix
  , parseOpts
  )
  where

import Control.Newtype.Generics (Newtype, op)
import GHC.Generics (Generic)
import Options.Applicative ( (<**>), auto, header, help, info, long, metavar,
                             option, showDefault, strOption, value, execParser,
                             helper, Parser )
import Text.Printf (printf)

newtype TitleLabel = TitleLabel String
  deriving (Generic, Show)

instance Newtype TitleLabel

newtype MarloweScan = MarloweScan Int
  deriving (Generic, Show)

instance Newtype MarloweScan

newtype RuntimeHost = RuntimeHost String
  deriving (Generic, Show)

instance Newtype RuntimeHost

newtype RuntimePort = RuntimePort Int
  deriving (Generic, Show)

instance Newtype RuntimePort

newtype BlockExplorerHost = BlockExplorerHost String
  deriving (Generic, Show)

instance Newtype BlockExplorerHost

data Options = Options
  { optTitleLabel :: TitleLabel
  , optMarloweScanPort :: MarloweScan
  , optRuntimeHost :: RuntimeHost
  , optRuntimePort :: RuntimePort
  , optBlockExplorerHost :: BlockExplorerHost
  }
  deriving Show

parser :: Parser Options
parser = Options
  <$> ( TitleLabel <$> strOption
        (  long "title-label"
        <> metavar "TEXT"
        <> help ("Label to be shown together with the title in MarloweScan in parenthesis. " ++
                 "(It can be used to display the name of the network that MarloweScan is deployed to.)")
        <> showDefault
        <> value "Preprod"
        )
      )
  <*> ( MarloweScan <$> option auto
        (  long "marlowe-scan-port"
        <> metavar "PORT"
        <> help "Port number to use for this MarloweScan server"
        <> showDefault
        <> value 8081
        )
      )
  <*> ( RuntimeHost <$> strOption
        (  long "runtime-host"
        <> metavar "HOSTNAME-OR-IP"
        <> help "Hostname or IP of the running Marlowe Runtime server"
        <> showDefault
        <> value "builder"
        )
      )
  <*> ( RuntimePort <$> option auto
        (  long "runtime-port"
        <> metavar "PORT"
        <> help "Port number of the running Marlowe Runtime server"
        <> showDefault
        <> value 8080
        )
      )
  <*> ( BlockExplorerHost <$> strOption
        (  long "block-explorer"
        <> metavar "HOST"
        <> help "Host for exploring Cardano blockchain addresses, transactions, etc."
        <> showDefault
        <> value "preprod.cardanoscan.io"
        )
      )

parseOpts :: IO Options
parseOpts = do
  execParser $ info (parser <**> helper)
    (  header "MarloweScan server"
    )

mkUrlPrefix :: Options -> String
mkUrlPrefix opts =
  let
    rhost = op RuntimeHost . optRuntimeHost $ opts
    rport = op RuntimePort . optRuntimePort $ opts
  in
    printf "http://%s:%d/" rhost rport
